src/VecSim/vec_sim_index.h

```diff
@@ -167,19 +169,23 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
             .fieldType = INFOFIELD_STRING,
             .fieldValue = {FieldValue{.stringValue = VecSimSearchMode_ToString(info.lastMode)}}});
     }
-    const void *processBlob(const void *original_blob, void *processed_blob) const {
+    const void *processBlob(const void *original_blob, void *aligned_mem) const {
+        void *processed_blob;
+        // if the blob is not aligned, or we need to normalize, we copy it
+        if ((uintptr_t)original_blob % this->alignment || this->metric == VecSimMetric_Cosine) {
+            memcpy(aligned_mem, original_blob, this->dataSize);
+            processed_blob = aligned_mem;
+        } else {
+            processed_blob = (void *)original_blob;
+        }
+
         // if the metric is cosine, we need to normalize
         if (this->metric == VecSimMetric_Cosine) {
-            // copy original blob to the output blob
-            memcpy(processed_blob, original_blob, this->dataSize);
             // normalize the copy in place
             normalize_func(processed_blob, this->dim);
```

Comment:
can be moved inside the first if to avoid checking if cosine twice when it isnt

Reply:
I trust the compiler to do so. This is done once in a query/insertion anyway

Reply:
in gcc we trust